### PR TITLE
decoupled result set from execution plan

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR decouples result set from execution plan building. This is enabler for #907

### CODE REVIEW

File: src/execution_plan/execution_plan.c [hunk 1]
```diff
@@ -605,8 +603,8 @@ static void _buildMergeOp(GraphContext *gc, AST *ast, ExecutionPlan *plan,
 	AST_MergeContext merge_ctx = AST_PrepareMergeOp(clause, plan->query_graph, bound_vars);
 
 	// Create a Merge operation. It will store no information at this time except for any graph updates
-	// it should make due to ON MATCH and ON CREATE SET directives in the query.
-	OpBase *merge_op = NewMergeOp(plan, merge_ctx.on_match, merge_ctx.on_create, stats);
+	// it should make due to ON MATCH SET directives in the query.
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: This comment shouldn't be changed.

File: src/execution_plan/execution_plan.c [hunk 2]
```diff
@@ -742,10 +733,10 @@ ExecutionPlan *ExecutionPlan_UnionPlans(ResultSet *result_set, AST *ast) {
 	plan->segment_count = union_count;
 	plan->query_graph = NULL;
 	plan->record_map = raxNew();
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: `plan->is_union = true;`

File: src/execution_plan/execution_plan.c [hunk 3]
```diff
@@ -789,13 +777,13 @@ static OpBase *_ExecutionPlan_FindLastWriter(OpBase *root) {
...
 
-ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
+ExecutionPlan *NewExecutionPlan() {
 	AST *ast = QueryCtx_GetAST();
 	uint clause_count = cypher_ast_query_nclauses(ast->root);
 
 	/* Handle UNION if there are any. */
 	if(AST_ContainsClause(ast, CYPHER_AST_UNION)) {
-		return ExecutionPlan_UnionPlans(result_set, ast);
+		return ExecutionPlan_UnionPlans(ast);
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: One-liner?

File: src/execution_plan/execution_plan.c [hunk 4]
```diff
@@ -994,25 +987,22 @@ void ExecutionPlan_Init(ExecutionPlan *plan) {
...
+void ExecutionPlan_Execute(ExecutionPlan *plan) {
 	/* Set an exception-handling breakpoint to capture run-time errors.
 	 * encountered_error will be set to 0 when setjmp is invoked, and will be nonzero if
 	 * a downstream exception returns us to this breakpoint. */
 	int encountered_error = SET_EXCEPTION_HANDLER();
 
 	if(encountered_error) {
 		// Encountered a run-time error; return immediately.
-		return plan->result_set;
+		return;
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: One-liner?

File: src/execution_plan/execution_plan.c [hunk 5]
```diff
@@ -885,39 +873,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
...
 	plan->segments = segments;
-	QueryCtx_SetLastWriter(_ExecutionPlan_FindLastWriter(plan->root));
+
 	return plan;
 }
 
+void ExecutionPlan_PreparePlan(ExecutionPlan *plan) {
+	optimizePlan(plan);
+	QueryCtx_SetLastWriter(_ExecutionPlan_FindLastWriter(plan->root));
+	if(plan->returned_column_names) ResultSet_SetColumns(QueryCtx_GetResultSet(),
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: Let's have `ResultSet_SetColumns` invoke `QueryCtx_GetResultSet` rather than expecting it as an argument.

File: src/execution_plan/execution_plan.c [hunk 6]
```diff
@@ -885,39 +873,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
...
+															 ExecutionPlan_GetResultColumns(plan));
+}
+
 inline rax *ExecutionPlan_GetMappings(const ExecutionPlan *plan) {
 	assert(plan && plan->record_map);
 	return plan->record_map;
 }
 
+inline const char **ExecutionPlan_GetResultColumns(const ExecutionPlan *plan) {
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: This function seems unnecessary to me - it's only invoked at line 901.

File: src/execution_plan/execution_plan.h [hunk 7]
```diff
@@ -17,14 +16,15 @@ typedef struct ExecutionPlan ExecutionPlan;
...
 	rax *record_map;                    // Mapping between identifiers and record indices.
-	ResultSet *result_set;              // ResultSet populated by this query.
 	QueryGraph *query_graph;            // QueryGraph representing all graph entities in this segment.
 	FT_FilterNode *filter_tree;         // FilterTree containing filters to be applied to this segment.
 	QueryGraph **connected_components;  // Array of all connected components in this segment.
 	// NOTE - segments and segment_count are only stored for proper freeing.
 	int segment_count;                  // Number of ExecutionPlan segments.
 	ExecutionPlan **segments;           // Partial execution plans scoped to a subset of operations.
 	ObjectPool *record_pool;
+	const char **returned_column_names; // The final result set column names.
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.h]: I think that `column_names` would be a fine name; it's not like there are other columns that aren't being returned.

File: src/execution_plan/execution_plan.h [hunk 8]
```diff
@@ -17,14 +16,15 @@ typedef struct ExecutionPlan ExecutionPlan;
...
+	bool is_union;                      // Indicates if the execution plan as a union of execution plans.
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.h]: This is currently always false, though that should be fixed by the change at 735. Do ExecutionPlans actually need to be aware of this at all, though? It looks like this is used as a switch for `optimizePlan`, but we can probably base the same switch on an AST check for a UNION clause - thoughts?

File: src/execution_plan/execution_plan.h [hunk 9]
```diff
@@ -88,13 +88,16 @@ OpBase *ExecutionPlan_BuildOpsFromPath(ExecutionPlan *plan, const char **vars,
 /* execution_plan.c */
 
 /* Creates a new execution plan from AST */
-ExecutionPlan *NewExecutionPlan(ResultSet *result_set);
+ExecutionPlan *NewExecutionPlan();
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.h]: `ExecutionPlan *NewExecutionPlan(void);`

File: src/execution_plan/execution_plan.h [hunk 10]
```diff
@@ -88,13 +88,16 @@ OpBase *ExecutionPlan_BuildOpsFromPath(ExecutionPlan *plan, const char **vars,
...
+
+/* Prepare an execution plan for execution: optimize, initialize result set schema. */
+void ExecutionPlan_PreparePlan(ExecutionPlan *plan);
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.h]: I think a name like `ExecutionPlan_FinalizePlan` might be better in order to more clearly convey that we're between building and executing the plan.

File: src/execution_plan/execution_plan.h [hunk 11]
```diff
@@ -107,6 +110,9 @@ void ExecutionPlan_PlaceFilterOps(ExecutionPlan *plan, const OpBase *recurse_lim
 /* Retrieve the map of aliases to Record offsets in this ExecutionPlan segment. */
 rax *ExecutionPlan_GetMappings(const ExecutionPlan *plan);
 
+/* Retrieve the result set column names. */
+const char **ExecutionPlan_GetResultColumns(const ExecutionPlan *plan);
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.h]: Even if you wish to keep this function, it doesn't need to be exposed in the header.

File: src/resultset/resultset.c [hunk 12]
```diff
@@ -128,8 +128,12 @@ ResultSet *NewResultSet(RedisModuleCtx *ctx, bool compact) {
 }
 
 void ResultSet_SetColumns(ResultSet *set, const char **columns) {
-	assert(set && columns);
-	set->columns = columns;
+	// Avoid situations like in EXPLAIN which has no result set.
+	if(!set) return;
+	// assert(columns);
```
Comment (jeffreylovitz) [inline src/resultset/resultset.c]: Delete commented assertion.

File: src/resultset/resultset.h [hunk 13]
```diff
@@ -36,6 +36,8 @@ void ResultSet_SetColumns(ResultSet *set, const char **columns);
 
 int ResultSet_AddRecord(ResultSet *set, Record r);
 
+void ResultSet_SetColumns(ResultSet *set, const char **column_names);
```
Comment (jeffreylovitz) [inline src/resultset/resultset.h]: Duplicate of line 35.

File: src/commands/cmd_query.c [hunk 14]
```diff
@@ -96,7 +96,7 @@ void Graph_Query(void *args) {
 	QueryCtx_SetResultSet(result_set);
```
Comment (jeffreylovitz) [inline src/commands/cmd_query.c]: IMO we may as well move the `QueryCtx_SetResultSet` call into `NewResultSet`, then have no local ResultSet variable at all in the various commands.

File: src/commands/cmd_profile.c [hunk 15]
```diff
@@ -61,7 +61,7 @@ void Graph_Profile(void *args) {
 	}
 
 	result_set = NewResultSet(ctx, false);
```
Comment (jeffreylovitz) [inline src/commands/cmd_profile.c]: If not making the change suggested [here](https://github.com/RedisGraph/RedisGraph/pull/929/files?utf8=%E2%9C%93&diff=split&w=1#r382827623), `QueryCtx_SetResultSet` needs to be called.

File: src/execution_plan/execution_plan.c (hunk 2 above)
Comment (DvirDukhan) [reply src/execution_plan/execution_plan.c]: i wonder why it was removed after rebase, and moreover, how the tests passed without it. I leave it for now, but it might be redundant.

File: src/execution_plan/execution_plan.h (hunk 8 above)
Comment (DvirDukhan) [reply src/execution_plan/execution_plan.h]: removed

File: src/execution_plan/execution_plan.h (hunk 10 above)
Comment (DvirDukhan) [reply src/execution_plan/execution_plan.h]: IMO, Finalize has different semantics, such as resource and memory free.

File: src/execution_plan/execution_plan.h (hunk 11 above)
Comment (DvirDukhan) [reply src/execution_plan/execution_plan.h]: removed

File: src/execution_plan/execution_plan.h [hunk 16]
```diff
@@ -17,14 +16,14 @@ typedef struct ExecutionPlan ExecutionPlan;
...
 	rax *record_map;                    // Mapping between identifiers and record indices.
-	ResultSet *result_set;              // ResultSet populated by this query.
 	QueryGraph *query_graph;            // QueryGraph representing all graph entities in this segment.
 	FT_FilterNode *filter_tree;         // FilterTree containing filters to be applied to this segment.
 	QueryGraph **connected_components;  // Array of all connected components in this segment.
 	// NOTE - segments and segment_count are only stored for proper freeing.
 	int segment_count;                  // Number of ExecutionPlan segments.
 	ExecutionPlan **segments;           // Partial execution plans scoped to a subset of operations.
 	ObjectPool *record_pool;
+	const char **column_names;          // The final result set column names.
```
Comment (DvirDukhan) [inline src/execution_plan/execution_plan.h]: remove and inject column names from AST

File: src/execution_plan/execution_plan.h [hunk 17]
```diff
@@ -88,13 +87,16 @@ OpBase *ExecutionPlan_BuildOpsFromPath(ExecutionPlan *plan, const char **vars,
 /* execution_plan.c */
 
 /* Creates a new execution plan from AST */
-ExecutionPlan *NewExecutionPlan(ResultSet *result_set);
+ExecutionPlan *NewExecutionPlan(void);
+
+/* Prepare an execution plan for execution: optimize, initialize result set schema. */
+void ExecutionPlan_PreparePlan(ExecutionPlan *plan);
```
Comment (DvirDukhan) [inline src/execution_plan/execution_plan.h]: add a flag for prepared that will assert upon execution

File: src/execution_plan/execution_plan.h [hunk 18]
```diff
@@ -120,10 +122,10 @@ void ExecutionPlan_Print(const ExecutionPlan *plan, RedisModuleCtx *ctx);
 void ExecutionPlan_Init(ExecutionPlan *plan);
 
 /* Executes plan */
-ResultSet *ExecutionPlan_Execute(ExecutionPlan *plan);
+void ExecutionPlan_Execute(ExecutionPlan *plan);
 
 /* Profile executes plan */
-ResultSet *ExecutionPlan_Profile(ExecutionPlan *plan);
+void ExecutionPlan_Profile(ExecutionPlan *plan);
```
Comment (DvirDukhan) [inline src/execution_plan/execution_plan.h]: return resultset

File: src/resultset/resultset.c [hunk 19]
```diff
@@ -206,9 +210,8 @@ void ResultSet_ReportQueryRuntime(RedisModuleCtx *ctx) {
 void ResultSet_Free(ResultSet *set) {
 	if(!set) return;
 
-	array_free(set->columns);
+	if(set->columns) array_free(set->columns);
 	rm_free(set->columns_record_map);
```
Comment (swilly22) [inline src/resultset/resultset.c]: Similar to the change at line 213, add `if(set->columns_record_map)`

File: src/resultset/resultset.c [hunk 20]
```diff
@@ -127,9 +127,13 @@ ResultSet *NewResultSet(RedisModuleCtx *ctx, bool compact) {
 	return set;
 }
 
-void ResultSet_SetColumns(ResultSet *set, const char **columns) {
-	assert(set && columns);
-	set->columns = columns;
+void ResultSet_SetColumns(const char **columns) {
+	ResultSet *set = QueryCtx_GetResultSet();
```
Comment (swilly22) [inline src/resultset/resultset.c]: I don't like this design where we call `ResultSet_*` functions without passing a result-set object,
please revert.

File: src/execution_plan/optimizations/optimizer.c [hunk 21]
```diff
@@ -48,3 +48,13 @@ void optimizePlan(ExecutionPlan *plan) {
...
 
+void optimizePlan(ExecutionPlan *plan) {
+	AST *ast = QueryCtx_GetAST();
+
+	/* Handle UNION if there are any. */
+	if(AST_ContainsClause(ast, CYPHER_AST_UNION)) {
+		for(uint i = 0; i < plan->segment_count; i++) _optimizePlan(plan->segments[i]);
+	} else {
+		_optimizePlan(plan);
+	}
```
Comment (swilly22) [inline src/execution_plan/optimizations/optimizer.c]: Does `plan->segments[0]` points to `plan` by any chance?
if so all we need is: `for(uint i = 0; i < plan->segment_count; i++) _optimizePlan(plan->segments[i]);`

Comment (swilly22) [reply src/execution_plan/optimizations/optimizer.c]: Also are there other situations where `plan->segment_count > 0` except UNION?
How would you define an execution plan segment?

File: src/execution_plan/optimizations/optimizer.c [hunk 22]
```diff
@@ -6,10 +6,10 @@
 
 #include "./optimizer.h"
 #include "./optimizations.h"
+#include "../../query_ctx.h"
 
-void optimizePlan(ExecutionPlan *plan) {
-
-	// Tries to compact filter trees, and remove redundant filters.
+void _optimizePlan(ExecutionPlan *plan) {
+// Tries to compact filter trees, and remove redundant filters.
```
Comment (swilly22) [inline src/execution_plan/optimizations/optimizer.c]: Ident.

File: src/execution_plan/execution_plan.h [hunk 23]
```diff
@@ -17,14 +17,14 @@ typedef struct ExecutionPlan ExecutionPlan;
...
 	rax *record_map;                    // Mapping between identifiers and record indices.
-	ResultSet *result_set;              // ResultSet populated by this query.
 	QueryGraph *query_graph;            // QueryGraph representing all graph entities in this segment.
 	FT_FilterNode *filter_tree;         // FilterTree containing filters to be applied to this segment.
 	QueryGraph **connected_components;  // Array of all connected components in this segment.
 	// NOTE - segments and segment_count are only stored for proper freeing.
 	int segment_count;                  // Number of ExecutionPlan segments.
 	ExecutionPlan **segments;           // Partial execution plans scoped to a subset of operations.
 	ObjectPool *record_pool;
+	bool is_exec_ready;                 // Indicates if the execution plan is ready for execute.
```
Comment (swilly22) [inline src/execution_plan/execution_plan.h]: Rename to `bool prepared`

File: src/commands/cmd_query.c (hunk 14 above)
Comment (swilly22) [reply src/commands/cmd_query.c]: I prefer to avoid this "hiding" I think being explicit here is preferable, 
`NewResultSet` should return a `result-set` object, let the caller decide what to do with it.

File: src/execution_plan/execution_plan.c (hunk 2 above)
Comment (swilly22) [reply src/execution_plan/execution_plan.c]: I think `is_union` is valuable, we can use it in the optimiser phase to quickly understand if we need to apply optimisation on the different segments.

File: src/execution_plan/execution_plan.c [hunk 24]
```diff
@@ -875,42 +860,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
...
-	if(query_has_return) {
-		if(!connecting_op) {
-			// Set the connecting op if our query is just a RETURN.
-			assert(segment_count == 1);
-			connecting_op = ExecutionPlan_LocateFirstOp(plan->root, OPType_PROJECT | OPType_AGGREGATE);
-		}
+	if(query_has_return || last_clause_type == CYPHER_AST_CALL) {
+		OpBase *results_op = NewResultsOp(plan);
+		_ExecutionPlan_UpdateRoot(plan, results_op);
+	}
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c]: Please verify `CALL proc UNION CALL proc`

File: src/execution_plan/execution_plan.c [hunk 25]
```diff
@@ -875,42 +860,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
...
+	const cypher_astnode_t *last_clause = cypher_ast_query_get_clause(ast->root, clause_count - 1);
+	cypher_astnode_type_t last_clause_type = cypher_astnode_type(last_clause);
+	bool query_has_return = (last_clause_type == CYPHER_AST_RETURN);
+	if(query_has_return) {
+		ResultSet_SetColumns(AST_BuildColumnNames(last_clause));
 	} else if(last_clause_type == CYPHER_AST_CALL) {
-		assert(plan->root->type == OPType_PROC_CALL);
-		OpProcCall *last_op = (OpProcCall *)plan->root;
+		OpBase *proc_call_base = ExecutionPlan_LocateFirstOp(plan->root, OPType_PROC_CALL);
+		assert(proc_call_base && (proc_call_base->type == OPType_PROC_CALL));
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c]: `proc_call_base->type == OPType_PROC_CALL` redundant, you've asked for: `ExecutionPlan_LocateFirstOp(plan->root, OPType_PROC_CALL);`

File: src/execution_plan/execution_plan.c [hunk 26]
```diff
@@ -875,42 +860,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
...
+	// Disregard self.
+	plan->segment_count = segment_count - 1;
+	plan->segments = segments;
 
-		OpBase *results_op = NewResultsOp(plan, result_set);
-		_ExecutionPlan_UpdateRoot(plan, results_op);
+	return plan;
+}
+
+static void _ExecutionPlan_SetResultColumns(ExecutionPlan *plan) {
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c]: It seems like this logic is a bit miss placed,
To me it makes more sense to place setting of the result-set columns within `ResultSet_New`.

File: src/execution_plan/execution_plan.c [hunk 27]
```diff
@@ -875,42 +860,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
...
-		_ExecutionPlan_UpdateRoot(plan, results_op);
+		const char **columns;
+		array_clone(columns, proc_call->output);
+		ResultSet_SetColumns(columns);
 	}
+}
 
-	// Optimize the operations in the ExecutionPlan.
+void ExecutionPlan_PreparePlan(ExecutionPlan *plan) {
+	_ExecutionPlan_SetResultColumns(plan);
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c]: I hope we'll be able to avoid this call, see my comment above.

File: src/execution_plan/execution_plan.c [hunk 28]
```diff
@@ -1000,19 +987,17 @@ ResultSet *ExecutionPlan_Execute(ExecutionPlan *plan) {
...
 	ExecutionPlan_Init(plan);
 
 	Record r = NULL;
 	// Execute the root operation and free the processed Record until the data stream is depleted.
 	while((r = OpBase_Consume(plan->root)) != NULL) ExecutionPlan_ReturnRecord(r->owner, r);
 
-	// Return the result set.
-	return plan->result_set;
+clean_up:
+	return QueryCtx_GetResultSet();
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c]: There's no real clean_up logic here, maybe we should revert this change.

File: src/execution_plan/execution_plan.c [hunk 29]
```diff
@@ -1041,11 +1026,10 @@ static void _ExecutionPlan_FinalizeProfiling(OpBase *root) {
...
 }
 
-ResultSet *ExecutionPlan_Profile(ExecutionPlan *plan) {
+void ExecutionPlan_Profile(ExecutionPlan *plan) {
 	_ExecutionPlan_InitProfiling(plan->root);
-	ResultSet *rs = ExecutionPlan_Execute(plan);
+	ExecutionPlan_Execute(plan);
 	_ExecutionPlan_FinalizeProfiling(plan->root);
-	return rs;
 }
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c]: Similar to `ExecutionPlan_Execute` `ExecutionPlan_Profile` should return a result-set.
the caller is expected to free it.

File: src/execution_plan/optimizations/optimizer.c (hunk 21 above)
Comment (DvirDukhan) [reply src/execution_plan/optimizations/optimizer.c]: in `UNION` situations, `segments` do not hold the plan itself
in regular situations, the plan is in index `segments_count +1`


from `NewExecutionPlan`
```
for(int i = 0; i < segment_count; i++) {
		uint end_offset = segment_indices[i];
		// Slice the AST to only include the clauses in the current segment.
		AST *ast_segment = AST_NewSegment(ast, start_offset, end_offset);
		ast_segments[i] = ast_segment;
		// Construct a new ExecutionPlanSegment.
		ExecutionPlan *segment = ExecutionPlan_NewEmptyExecutionPlan();
		ExecutionPlan_PopulateExecutionPlan(segment);

		segments[i] = segment;
		start_offset = end_offset;
	}
```
and
```
	// Disregard self.
	plan->segment_count = segment_count - 1;
	plan->segments = segments;
```
